src/native/libs/System.Globalization.Native/pal_collation.m

```diff
@@ -279,4 +280,82 @@ int32_t GlobalizationNative_EndsWithNative(const uint16_t* localeName, int32_t l
     return result == NSOrderedSame ? 1 : 0;
 }
 
+/**
+ * Append a code point to a string, overwriting 1 or 2 code units.
+ * The offset points to the current end of the string contents
+ * and is advanced (post-increment).
+ * "Safe" macro, checks for a valid code point.
+ * If a surrogate pair is written, checks for sufficient space in the string.
+ * If the code point is not valid or a trail surrogate does not fit,
+ * then isError is set to true.
+ *
+ * @param buffer const uint16_t * string buffer
+ * @param offset string offset, must be offset<capacity
+ * @param capacity size of the string buffer
+ * @param codePoint code point to append
+ * @param isError output bool set to true if an error occurs, otherwise not modified
+ */
+#define Append(buffer, offset, capacity, codePoint, isError) { \
+    if ((uint32_t)(codePoint) <= 0xffff) { \
+        (buffer)[(offset)++] = (uint16_t)(codePoint); \
+    } else if ((uint32_t)(codePoint) <= 0x10ffff && (offset) + 1 < (capacity)) { \
+        (buffer)[(offset)++] = (uint16_t)(((codePoint) >> 10) + 0xd7c0); \
+        (buffer)[(offset)++] = (uint16_t)(((codePoint)&0x3ff) | 0xdc00); \
+    } else /* codePoint > 0x10ffff or not enough space */ { \
+        (isError) = true; \
+    } \
+}
+
+/*
+Function:
+ChangeCaseNative
+
+Returns upper or lower casing of a string, taking into account the specified locale.
+*/
+int32_t GlobalizationNative_ChangeCaseNative(const uint16_t* localeName, int32_t lNameLength,
+                                             const uint16_t* lpSrc, int32_t cwSrcLength, uint16_t* lpDst, int32_t cwDstLength, int32_t bToUpper)
+{
+    NSLocale *currentLocale = GetCurrentLocale(localeName, lNameLength);
+    NSString *source = [NSString stringWithCharacters: lpSrc length: cwSrcLength];
+    NSString *result = bToUpper ? [source uppercaseStringWithLocale:currentLocale] : [source lowercaseStringWithLocale:currentLocale];
+
+    int32_t srcIdx = 0, dstIdx = 0, isError = false;
+    uint16_t dstCodepoint;
+    if (result.length > cwDstLength)
+        result = source;
```

Comment:
Why are we doing this (this is probably why `yield return new object[] { cultureName, "\u00DF", "\u00DF" };` isnt failing)? Wouldn't this mean that the resulting `lpDst` is not correctly converted to uppercase/lowercase?

Instead maybe we should indicate that the destination buffer isn't large enough.

Reply:
We are doing this to keep current logic with ICU which is hanging case uses simple folding: doesn't change UTF-16 code unit count.

Reply:
I couldn't find an equivalent logic for how current ICU prevents the UTF-16 code unit count from changing besides just having [the while loop condition](https://github.com/dotnet/runtime/blob/5ca682626651957080b6b4c42cb2ee1c5e743eb3/src/native/libs/System.Globalization.Native/pal_casing.c#L45C9-L45C37).

Wouldn't that suffice for retaining the original count? It feels like setting the string back to source instead of the capitalized/lower cased version handles this case differently from the original ICU logic.

Reply:
My bad, I made assumption from loop condition and some test cases. 
Updated that part and added behavioural changes in the doc.